### PR TITLE
Quartz: auto delete jobs without a class during scheduler initialization

### LINKED ISSUE #36184: Delete Quartz jobs & triggers if their job class doesn't exist
When a persistent Quartz job is removed or renamed, the job and trigger definitions stick around in the App DB, and print long exceptions in the logs when they fire because the associated class for the job doesn't exist in the codebase. This isn't a huge issue aside from the log spam, but it would be nice if we could modify Quartz jobs without worrying about having to clean up after ourselves.

In https://github.com/metabase/metabase/pull/35940 I'm catching these exceptions and logging a shorter warning (without the long stack trace) to avoid polluting the logs too much.

We should also consider deleting jobs & triggers automatically if their class can't be found. If there's no class, the job can't run, so I think it's safe to use this as an indication that we should clean it up.

[Related Slack thread](https://metaboat.slack.com/archives/CKZEMT1MJ/p1701192612446319)

### PR DESCRIPTION
**Pull request by qnkhuat**
Closes https://github.com/metabase/metabase/issues/36184

Whenever we delete a Quartz Job we have to write a custom migration to delete it, see `DeleteTruncateAuditLogTask` for example. 

This makes it so on initialization of task scheduler, we will automatically delete any jobs that can't find a class.

I don't know how to write a test that involves making a class disappear, but any ideas are welcome!

But here is how I manually tested it.
1. Start MB normally
2. Comment out the `(delete-jobs-with-no-class!)` call in `init-scheduler!`
3. Comment out all the code in `metabase.task.follow-up-emails` to make the `FollowUpEmail` disappear
4. Restart MB, eval `(metabase.task/scheduler-info)` in the reply will print out an exception saying `FollowUpEmail` not found
5. Uncomment `(delete-jobs-with-no-class!)` in step 2, and restart MB
6. Eval `(metabase.task/scheduler-info)`, this time there should be no errors because the `metabase.task.follow-up-emails.job` has been deleted.

This will also prevent us from having situations like this: https://github.com/metabase/metabase/pull/41772#issuecomment-2100263983

### CODE REVIEW

File: src/metabase/task.clj [hunk 1]
```diff
@@ -141,6 +141,18 @@
...
+  "Delete any jobs that have been scheduled but whose class is no longer available."
+  []
+  (when-let [scheduler (scheduler)]
+    (doseq [job-key (.getJobKeys scheduler nil)]
+      (try
+        (qs/get-job scheduler job-key)
+        (catch JobPersistenceException e
+          (when (instance? ClassNotFoundException (.getCause e))
+            (log/infof "Deleting job %s due to class not found" (.getName ^JobKey job-key))
+            (qs/delete-job scheduler job-key)))))))
```
Comment (calherries) [inline src/metabase/task.clj:154]: Do we also need to delete the triggers?

**Review by calherries — COMMENTED** [review]
Nice.

It might feel a little bit strange, but I think a test that relies on the implementation details of quartz could give us the safety we want:
- create a job detail and a trigger instance, similar to [this](https://github.com/metabase/metabase/blob/936c185a874ff25a7ecc4aa69947e3902a058cf4/test/metabase/task_test.clj#L63-L69)
- modify the `job_class_name` column in the `qrtz_job_details` table to a non-existent class name.
- restart the scheduler using `start-scheduler!`
- check that the job detail and trigger instances do not exist anymore

The test might be brittle if Quartz changes how they persist job class names but that's probably unlikely. The key thing is to know the test can't pass with a false positive result.

**Review by calherries — APPROVED** [review]

File: src/metabase/task.clj (hunk 1 above)
Comment (qnkhuat) [reply src/metabase/task.clj:154]: deleting the job will also delete all of its triggers.

**Comment by crisptrutski** [comment]
+1 to Cal's idea for a test. That's much simpler than trying to fiddle with class loaders, which I'm not sure is even possible.

**Comment by crisptrutski** [comment]
I'm a bit conflicted on this change. While I understand the ergonomics, I dislike the asymmetry between how we add the triggers and remove them. I also dislike that if we changed our mind and brought back the class it would be non-deterministic whether it's removed depending on how many versions the customer is upgrading.

I'd lean towards making it more convenient to do the clean-up versus making it completely automatic.

**Comment by piranha** [comment]
> I also dislike that if we changed our mind and brought back the class it would be non-deterministic whether it's removed depending on how many versions the customer is upgrading.

On the other hand, all of our jobs have their triggers/schedules configured all the time and it feels like it should not be a problem? If it does not exist, it'll add itself?

So the opposite is - if it does not exist, it should be removed?

**Comment by crisptrutski** [comment]
> If it does not exist, it'll add itself?

I think I'm missing something, let's clarify in Slack 😄

**Review by crisptrutski — APPROVED** [review]
> I think I'm missing something

Yeah, I was just confused. For some reason I thought we were initializing triggers in migrations - hence the distinction I was making about rollbacks. I see the creation is fairly implicit so no qualms with destruction being the same.

**Comment by qnkhuat** [comment]
@crisptrutski @calherries added a test

**Pull request merged**